engine/executor/dm/worker.go

```diff
@@ -0,0 +1,241 @@
+	return newDMWorker(ctx, masterID, f.workerType, conf.(*dmconfig.SubTaskConfig)), nil
+}
+
+// dmWorker implements methods for lib.WorkerImpl
+type dmWorker struct {
+	lib.BaseWorker
+
+	unitHolder   unitHolder
+	messageAgent dmpkg.MessageAgent
+
+	mu                 sync.RWMutex
+	cfg                *dmconfig.SubTaskConfig
+	storageWriteHandle broker.Handle
+	stage              metadata.TaskStage
+	workerType         libModel.WorkerType
+	taskID             string
+	masterID           libModel.MasterID
+}
+
+func newDMWorker(ctx *dcontext.Context, masterID libModel.MasterID, workerType lib.WorkerType, cfg *dmconfig.SubTaskConfig) *dmWorker {
+	w := &dmWorker{
+		cfg:        cfg,
+		stage:      metadata.StageInit,
+		workerType: workerType,
+		taskID:     cfg.SourceID,
+		masterID:   masterID,
+		unitHolder: newUnitHolderImpl(workerType, cfg),
+	}
+
+	// nolint:errcheck
+	ctx.Deps().Construct(func(m p2p.MessageHandlerManager) (p2p.MessageHandlerManager, error) {
+		w.messageAgent = dmpkg.NewMessageAgentImpl(w.taskID, w, m)
+		return m, nil
+	})
+	return w
+}
+
+// InitImpl implements lib.WorkerImpl.InitImpl
+func (w *dmWorker) InitImpl(ctx context.Context) error {
+	if err := w.messageAgent.Init(ctx); err != nil {
+		return err
+	}
+	// register jobmaster sender
+	if err := w.messageAgent.UpdateSender(w.masterID, w); err != nil {
+		return err
+	}
+	if w.cfg.Mode != dmconfig.ModeIncrement {
+		if err := w.setupstorage(ctx); err != nil {
+			return err
+		}
+	}
+	return w.unitHolder.Init(ctx)
+}
+
+// Tick implements lib.WorkerImpl.Tick
+func (w *dmWorker) Tick(ctx context.Context) error {
+	if err := w.unitHolder.Tick(ctx); err != nil {
+		return err
+	}
+	if err := w.tryUpdateStatus(ctx); err != nil {
+		return err
+	}
+	return w.messageAgent.Tick(ctx)
+}
+
+// Workload implements lib.WorkerImpl.Worload
+func (w *dmWorker) Workload() model.RescUnit {
+	log.L().Info("dmworker.Workload")
+	return 0
+}
+
+// OnMasterFailover implements lib.WorkerImpl.OnMasterFailover
+func (w *dmWorker) OnMasterFailover(reason lib.MasterFailoverReason) error {
+	log.L().Info("dmworker.OnMasterFailover")
+	return nil
+}
+
+// OnMasterMessage implements lib.WorkerImpl.OnMasterMessage
+func (w *dmWorker) OnMasterMessage(topic p2p.Topic, message p2p.MessageValue) error {
+	log.L().Info("dmworker.OnMasterMessage", zap.String("topic", topic), zap.Any("message", message))
+	return nil
+}
+
+// CloseImpl implements lib.WorkerImpl.CloseImpl
+func (w *dmWorker) CloseImpl(ctx context.Context) error {
+	// unreigster jobmaster sender
+	if err := w.messageAgent.UpdateSender(w.masterID, nil); err != nil {
+		return err
+	}
+	w.unitHolder.Close(ctx)
+	w.messageAgent.Close(ctx)
+	return nil
+}
+
+// exit closes the task and exits.
+func (w *dmWorker) exit(ctx context.Context, status libModel.WorkerStatus, err error) error {
```

Comment:
I think it's a bit weird to let a private function call public function of same name (though Exit is provided by lib.BaseWorker). we can rename exit to closeAndExit.

Or the framework should move CloseImpl inside the Exit? @liuzix